presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MultiJoinNode.java

```diff
@@ -0,0 +1,141 @@
+        this.filter = requireNonNull(filter, "filter is null");
+        this.outputSymbols = ImmutableList.copyOf(requireNonNull(outputSymbols, "outputSymbols is null"));
+
+        List<Symbol> inputSymbols = sources.stream().flatMap(source -> source.getOutputSymbols().stream()).collect(toImmutableList());
+        checkArgument(inputSymbols.containsAll(outputSymbols), "inputs do not contain all output symbols");
+    }
+
+    public Expression getFilter()
+    {
+        return filter;
+    }
+
+    public List<PlanNode> getSources()
+    {
+        return sources;
+    }
+
+    public List<Symbol> getOutputSymbols()
+    {
+        return outputSymbols;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(
+                ImmutableSet.copyOf(this.getOutputSymbols()),
+                ImmutableSet.copyOf(extractConjuncts(this.getFilter())),
+                this.getSources().stream()
+                        .map(PlanNode::getId)
+                        .collect(toImmutableSet()));
+    }
+
+    @Override
+    public boolean equals(Object other)
+    {
+        if (!(other instanceof MultiJoinNode)) {
+            return false;
+        }
+
+        MultiJoinNode otherNode = (MultiJoinNode) other;
+        return ImmutableSet.copyOf(this.getOutputSymbols()).equals(ImmutableSet.copyOf(otherNode.getOutputSymbols()))
+                && ImmutableSet.copyOf(extractConjuncts(this.getFilter())).equals(ImmutableSet.copyOf(extractConjuncts(otherNode.getFilter())))
+                && this.getSources().stream().map(PlanNode::getId).collect(toImmutableSet()).equals(otherNode.getSources().stream().map(PlanNode::getId).collect(toImmutableSet()));
+    }
+
+    static MultiJoinNode toMultiJoinNode(JoinNode joinNode, Lookup lookup)
+    {
+        return new MultiJoinNodeBuilder(joinNode, lookup).toMultiJoinNode();
+    }
+
+    private static class MultiJoinNodeBuilder
+    {
+        private final List<PlanNode> sources = new ArrayList<>();
+        private final List<Expression> filters = new ArrayList<>();
+        private final List<Symbol> outputSymbols;
+        private final Lookup lookup;
+
+        MultiJoinNodeBuilder(JoinNode node, Lookup lookup)
+        {
+            requireNonNull(node, "node is null");
+            checkState(node.getType() == INNER, "join type must be INNER");
+            this.outputSymbols = node.getOutputSymbols();
+            this.lookup = requireNonNull(lookup, "lookup is null");
+            flattenNode(node);
+        }
+
+        private void flattenNode(PlanNode node)
+        {
+            PlanNode resolved = lookup.resolve(node);
+            if (resolved instanceof JoinNode && ((JoinNode) resolved).getType() == INNER && isDeterministic(((JoinNode) resolved).getFilter().orElse(TRUE_LITERAL))) {
+                JoinNode joinNode = (JoinNode) resolved;
+                flattenNode(joinNode.getLeft());
```

Comment:
@ArturGajowy , i couldn't agree less :)
As things are now, you can't prioritize rule execution within iterative optimizer. In order to be able to construct `MultiJoinNode`, you'd need to have one iterative optimizer instance just for that and then another iterative optimizer to do join enumeration. The `MultiJoinNode` would need to be a proper `PlanNode` even though it's very transient in its nature. You'd need to have validator that checks it has indeed been removed from the plan. ... Maybe it would be more pure an approach, but I see the cost too high and, frankly, with the limitation of having dedicated iteratitve optimizers just for that, I don't see benefits.